### PR TITLE
[boost] Review C++ standard settings (fix C++11 on macOS builds)

### PR DESCRIPTION
**Pull request by david-sinuela-pix4d**
Specify library name and version:  **boost/***

While building for mac in our CI, some of the boost libraries were not compiled (fiber for example). After digging it resulted that C++11 was not enabled in the build.

This PR attempts to fix it by passing the `cxxstd=11` flag when building with apple-clang.

I've simplified a bit the code setting the cxx and link flags. It works fine for the platforms we are building Ubuntu18, Windows and macOS but it is very difficult for me to say if this will break anyone else's build. Can someone more familiar with this recipe tell which use cases should be supported?

The original build log showed this:

```
Building the Boost C++ Libraries.

    - C++11 mutex              : no
    - lockfree boost::atomic_flag : yes
    - Boost.Config Feature Check: cxx11_auto_declarations : no
    - Boost.Config Feature Check: cxx11_constexpr : no
    - Boost.Config Feature Check: cxx11_defaulted_functions : no
    - Boost.Config Feature Check: cxx11_final : no
    - Boost.Config Feature Check: cxx11_hdr_mutex : no
    - Boost.Config Feature Check: cxx11_hdr_tuple : no
    - Boost.Config Feature Check: cxx11_lambdas : no
    - Boost.Config Feature Check: cxx11_noexcept : no
    - Boost.Config Feature Check: cxx11_nullptr : no
    - Boost.Config Feature Check: cxx11_rvalue_references : no
    - Boost.Config Feature Check: cxx11_template_aliases : no
    - Boost.Config Feature Check: cxx11_thread_local : no
    - Boost.Config Feature Check: cxx11_variadic_templates : no
```

and the new build log shows:
```
Building the Boost C++ Libraries.

    - C++11 mutex              : yes
    - lockfree boost::atomic_flag : yes
    - Boost.Config Feature Check: cxx11_auto_declarations : yes
    - Boost.Config Feature Check: cxx11_constexpr : yes
    - Boost.Config Feature Check: cxx11_defaulted_functions : yes
    - Boost.Config Feature Check: cxx11_final : yes
    - Boost.Config Feature Check: cxx11_hdr_mutex : yes
    - Boost.Config Feature Check: cxx11_hdr_tuple : yes
    - Boost.Config Feature Check: cxx11_lambdas : yes
    - Boost.Config Feature Check: cxx11_noexcept : yes
    - Boost.Config Feature Check: cxx11_nullptr : yes
    - Boost.Config Feature Check: cxx11_rvalue_references : yes
    - Boost.Config Feature Check: cxx11_template_aliases : yes
    - Boost.Config Feature Check: cxx11_thread_local : yes
    - Boost.Config Feature Check: cxx11_variadic_templates : yes
```

### CODE REVIEW

File: recipes/boost/all/conanfile.py [hunk 1]
```diff
@@ -473,18 +472,15 @@ def _b2_abi(self):
             return None
 
     @property
-    def _gnu_cxx11_abi(self):
-        """Checks libcxx setting and returns value for the GNU C++11 ABI flag
-        _GLIBCXX_USE_CXX11_ABI= .  Returns None if C++ library cannot be
-        determined.
+    def _gnu_cxx11_abi_define(self):
```
Comment (david-sinuela-pix4d) [inline recipes/boost/all/conanfile.py]: This is a minor simplification.

File: recipes/boost/all/conanfile.py [hunk 2]
```diff
@@ -543,36 +539,6 @@ def add_defines(option, library):
         toolset, _, _ = self._get_toolset_version_and_exe()
         flags.append("toolset=%s" % toolset)
 
-        if self.settings.get_safe("compiler.cppstd"):
-            flags.append("cxxflags=%s" % cppstd_flag(
-                    self.settings.get_safe("compiler"),
-                    self.settings.get_safe("compiler.version"),
-                    self.settings.get_safe("compiler.cppstd")
-                )
-            )
```
Comment (david-sinuela-pix4d) [inline recipes/boost/all/conanfile.py]: This code did not work as expected, at least when using apple-clang. I'll post the conan profile used in the main comments.

**Comment by david-sinuela-pix4d** [comment]
This is the conan profile used for macos:

```
# ~/.conan/profiles/apple_clang
[settings]
os=Macos
os.version=10.14
os_build=Macos
arch=x86_64
arch_build=x86_64
compiler=apple-clang
compiler.version=10.0
compiler.libcxx=libc++
build_type=Release
[options]
[build_requires]
[env]
```

**Comment by david-sinuela-pix4d** [comment]
Ready for review, I restored the code setting the stdlib for clang.

File: recipes/boost/all/conanfile.py [hunk 3]
```diff
@@ -871,9 +860,8 @@ def package_info(self):
         if self.options.segmented_stacks:
             self.cpp_info.defines.extend(["BOOST_USE_SEGMENTED_STACKS", "BOOST_USE_UCONTEXT"])
 
-        if self.settings.os != "Android":
```
Comment (uilianries) [inline recipes/boost/all/conanfile.py]: How do you know if this change will not break Android? Did you test, do you have a log?

Comment (david-sinuela-pix4d) [reply recipes/boost/all/conanfile.py]: My understanding is that android can only compile with clang and libc++, that's why I believe the condition can be removed.

But, in fact I did not test it neither I have the resources to test it: android dev env, time, etc. My idea was to discover issues during the review, either through the CI or community review.

Could you or do you know who could test this recipe?

Comment (uilianries) [reply recipes/boost/all/conanfile.py]: you can use the Android NDK for testing: https://docs.conan.io/en/latest/integrations/cross_platform/android.html

**Comment by SSE4** [comment]
@david-sinuela-pix4d the correct way to enable **C++11** (or later) is by using **compiler.cppstd** setting. e.g. your profile may look like:
```
# ~/.conan/profiles/apple_clang
[settings]
os=Macos
os.version=10.14
os_build=Macos
arch=x86_64
arch_build=x86_64
compiler=apple-clang
compiler.version=10.0
compiler.libcxx=libc++
compiler.cppstd=11
build_type=Release
[options]
[build_requires]
[env]
```
notice the addition: **compiler.cppstd=11**
I suppose it should result in enabling of `cxxstd=11` and building additional libraries (e.g. **boost.fiber**)
please test if it works for you (without changing recipe) and lemme know.
I personally believe assuming clang is always C++11 standard is incorrect - it might be lower or upper standard, e.g. C++03 or C++20. we usually try to avoid such smart detection things in recipes and ask users to specify their settings explicitly (in profile or command line).

**Comment by david-sinuela-pix4d** [comment]
Back from holidays :)

Thanks @SSE4, I'll try that approach, less implicit behavior is clearly simpler to manage. If that is the better solution I will revert most changes and only contribute the changes to `_gnu_cxx11_abi_define`.

One question that raises to my mind is if the  *compiler.cppstd* setting is set to 11 or higher on apple-clang by default. The logs show that it is not when it is not set from the profile but does conan initialize the profile with cppstd=11 automatically or can the profile generation be changed to do so?

**Comment by david-sinuela-pix4d** [comment]
It fails with this error message if cppstd is defined in the profile:

```
ERROR: 'settings.compiler.cppstd' doesn't exist for 'apple-clang'
'settings.compiler' possible configurations are ['libcxx', 'version']
```

That was with conan 1.18.6 (the version we have in production). I tried with the latest version (1.21.0) and it reports the same error. Am I doing anything wrong or is `cppstd` not supported for `apple-clang`? I also re-tested the recipe without the changes with conan 1.21.0 and it didn't build with c++11 by default either.

**Comment by david-sinuela-pix4d** [comment]
I managed to set the `cppstd` setting by removing the `settings.yml` file in the conan folder and letting conan re-create it.

I performed a few more tests, these are the results:
 * Both conan 1.18.6 and 1.21.0 behave in the same way.
 * The default settings for `apple-clang` do not seem to define `compiler.cppstd`.
 * Defining `compiler.cppstd=11` in the profile enables c++ 11 features in the original boost recipe.
 * In order to use `compiler.cppstd` the conan settings file needs to be updated. An easy way to do so is removing the `settings.yml` file in conan's configuration folder and letting conan re-create it. This is a bit worrying because many users will need to update the profiles and settings file. This includes developers and CI systems.

Some general questions with an attempt to find what are the next steps:

1. I am not sure if not defining a default `cppstd` for apple-clang is intentional or a bug. The documentation seems to contradict itself in https://docs.conan.io/en/latest/howtos/manage_cpp_standard.html#how-to-manage-c-standard-experimental and https://docs.conan.io/en/latest/howtos/manage_cpp_standard.html#package-compatibility.
2. With `cppstd` labeled as experimental in the documentation, isn't it risky to depend on it?

**Comment by SSE4** [comment]
@david-sinuela-pix4d I believe the default configuration is C++03 for Apple Clang compilers, which is the same effect as not defining **cppstd** in profile. if you need another standard (14, 17 or 20), you'll have to define **cppstd** within profile and build from sources, as ConanCenterIndex will not build and upload such configurations.

**Comment by SSE4** [comment]
I believe no changes are needed to compile boost with C++11 support. the only change to be done is to define **compiler.cppstd** in conan profile. 
current PR changes many things that aren't intended:
- it removes **cppstd_flag** (so no **-std=c++11** or similar will be applied into compiler flags)
- it touches Android-specific conditions
- it always forces **cxxstd=11** for clang, making it impossible to build for older or newer standard with clang compiler

**Comment by SSE4** [comment]
@david-sinuela-pix4d please let me know if you still need changes in recipe, or this PR is unnecessary and defining `cppstd` in profile is enough to get C++11 compatible boost

**Comment by david-sinuela-pix4d** [comment]
@SSE4 sorry for the slow replies, after testing the solution proposed, (adding `compiler.cppstd` to the profile) works on our side, I may contribute some other minor changes later but this PR can be closed.

Thanks for the support.

**Pull request closed without merging**